### PR TITLE
CSHARP-4157: Add FLE 2 API to ClientEncryptionOpts.

### PR DESCRIPTION
**Pull request by DmitryLukyanov**

### CODE REVIEW

File: src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs [hunk 1]
```diff
@@ -104,7 +104,13 @@ private CryptOptions CreateCryptOptions()
                 encryptedFieldsBytes = GetBytesFromMap(_encryptedFieldsMap);
             }
 
-            return new CryptOptions(kmsProviders, encryptedFieldsMap: encryptedFieldsBytes, schema: schemaBytes, bypassQueryAnalysis: _bypassQueryAnalysis.GetValueOrDefault(false));
+            return new CryptOptions(
+                kmsProviders,
+                encryptedFieldsMap: encryptedFieldsBytes,
+                schema: schemaBytes,
+                bypassQueryAnalysis: _bypassQueryAnalysis.GetValueOrDefault(false),
+                csfleLibPath: "",
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Clusters/CryptClientCreator.cs]: temporary values

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs [hunk 2]
```diff
@@ -1398,7 +1530,11 @@ private void AssertInnerEncryptionException(Exception ex, Type exType, params st
                 extraOptions,
                 bypassAutoEncryption,
                 maxPoolSize);
-            autoEncryptionOptionsConfigurator?.Invoke(configuredSettings.AutoEncryptionOptions);
+
+            if (autoEncryptionOptionsConfigurator != null)
+            {
+                configuredSettings.AutoEncryptionOptions = autoEncryptionOptionsConfigurator?.Invoke(configuredSettings.AutoEncryptionOptions);
```
Comment (BorisDog) [inline tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: minor: it's either `autoEncryptionOptionsConfigurator != null` or `autoEncryptionOptionsConfigurator?`

**Review by BorisDog — APPROVED** [review]
LGTM (see minor comment)

File: tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs (hunk 2 above)
Comment (DmitryLukyanov) [reply tests/MongoDB.Driver.Tests/Specifications/client-side-encryption/prose-tests/ClientEncryptionProseTests.cs]: good point. This one actually sets options to null if configurator is null. Removed unnecessary `?` part

File: Docs/reference/content/reference/driver/admin.md [hunk 3]
```diff
@@ -64,7 +64,7 @@ using (var cursor = await client.ListDatabaseAsync())
 
 ## Collections
 
-These operations exists on the [`IMongoDatabase`]({{< apiref "T_MongoDB_Driver_IMongoDatabase" >}}) interface.
+These operations exist on the [`IMongoDatabase`]({{< apiref "T_MongoDB_Driver_IMongoDatabase" >}}) interface.
```
Comment (DmitryLukyanov) [inline Docs/reference/content/reference/driver/admin.md:67]: cherrypicked from master to make tests passed

File: src/MongoDB.Driver.Core/Core/Operations/CreateCollectionOperation.cs [hunk 4]
```diff
@@ -324,6 +325,18 @@ public WriteConcern WriteConcern
             set { _writeConcern = value; }
         }
 
+        /// <summary>
+        /// Gets or sets the clustered index definition.
+        /// </summary>
+        /// <value>
+        /// The clustered index definition.
+        /// </value>
+        public BsonDocument ClusteredIndex
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver.Core/Core/Operations/CreateCollectionOperation.cs:334]: cherrypicked from master to make tests passed

File: src/MongoDB.Driver/ClusteredIndexOptions.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+using MongoDB.Bson;
+using MongoDB.Bson.Serialization;
+
+namespace MongoDB.Driver
+{
+    /// <summary>
+    /// Options for creating a clustered index.
+    /// </summary>
+    public class ClusteredIndexOptions<TDocument>
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/ClusteredIndexOptions.cs:24]: cherrypicked from master to make tests passed

File: src/MongoDB.Driver/CreateCollectionOptions.cs [hunk 6]
```diff
@@ -237,10 +237,20 @@ internal static CreateCollectionOptions<TDocument> CoercedFrom(CreateCollectionO
...
         // private fields
+        private ClusteredIndexOptions<TDocument> _clusteredIndex;
         private IBsonSerializer<TDocument> _documentSerializer;
         private FilterDefinition<TDocument> _validator;
 
         // public properties
+        /// <summary>
+        /// Gets or sets the <see cref="ClusteredIndexOptions{TDocument}"/>.
+        /// </summary>
+        public ClusteredIndexOptions<TDocument> ClusteredIndex
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/CreateCollectionOptions.cs:248]: cherrypicked from master to make tests passed

File: src/MongoDB.Driver/MongoDatabaseImpl.cs [hunk 7]
```diff
@@ -216,34 +216,48 @@ public override void CreateCollection(IClientSessionHandle session, string name,
             var genericMethodDefinition = typeof(MongoDatabaseImpl).GetTypeInfo().GetMethod("CreateCollectionHelper", BindingFlags.NonPublic | BindingFlags.Instance);
             var documentType = options.GetType().GetTypeInfo().GetGenericArguments()[0];
             var methodInfo = genericMethodDefinition.MakeGenericMethod(documentType);
-            methodInfo.Invoke(this, new object[] { session, name, options, cancellationToken });
+            try
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/MongoDatabaseImpl.cs:219]: we have asserting on particular exception that is hidden over TargetInvocationException. This logic of course it's better to be rewritten on some day..

File: src/MongoDB.Driver/MongoDatabaseImpl.cs [hunk 8]
```diff
@@ -644,13 +658,11 @@ private Task CreateCollectionHelperAsync<TDocument>(IClientSessionHandle session
...
         private IWriteOperation<BsonDocument> CreateCreateCollectionOperation<TDocument>(string name, CreateCollectionOptions<TDocument> options)
         {
-            BsonDocument validator = null;
-            if (options.Validator != null)
-            {
-                var serializerRegistry = options.SerializerRegistry ?? BsonSerializer.SerializerRegistry;
-                var documentSerializer = options.DocumentSerializer ?? serializerRegistry.GetSerializer<TDocument>();
-                validator = options.Validator.Render(documentSerializer, serializerRegistry, _linqProvider);
-            }
+            var serializerRegistry = options.SerializerRegistry ?? BsonSerializer.SerializerRegistry;
```
Comment (DmitryLukyanov) [inline src/MongoDB.Driver/MongoDatabaseImpl.cs:663]: cherrypicked from master to make tests passed

**Pull request merged**